File: python/src/secrets.py
#!/usr/bin/env python

#############################################################################
# HEADER                                                                    #
#############################################################################
"""
encapsulation to secret management api using Vault
"""

#############################################################################
# IMPORT MODULES                                                            #
#############################################################################
import hvac
from .os_vars import OSVars

# TODO: move to a common folder within the config clients monorepo
VAULT_URL_KEY = "VAULT_URL"
VAULT_DEFAULT_URL = "https://vault.twistbioscience-staging.com"
VAULT_USER_KEY = "VAULT_USER"
VAULT_PASS_KEY = "VAULT_PASSWORD"

#############################################################################
# IMPLEMENTATION                                                            #
#############################################################################

OSVars.register_mandatory(VAULT_USER_KEY, "Vault secret management user name", str)
OSVars.register_mandatory(VAULT_PASS_KEY, "Vault secret management password", str)

OSVars.register(VAULT_URL_KEY, "Vault secret management server", str, VAULT_DEFAULT_URL)


class Secrets:
    """
    The Secret Management class serving application with credentials, keys and other secretive
    information that should not reside within the concealed configuration or environment variables.
    This is a read only implementation that can support dynamic environments, seperating dev, qa and staging
    secrets from production ones.

    Class Public Interface:

        connect
        get (already ensures call to connect)

    TODO: mem-encrypt cached elements for enhanced security
    TODO: add backward compatibility to old env secrets (qa | dev | staging)

    """

    __instance = None

    @staticmethod
    def instance():
        """ Static access method. """
        if Secrets.__instance is None:
            # the only place the Secrets constructor is being called.
            Secrets()
        return Secrets.__instance

    def __init__(self):
        if Secrets.__instance is None:
            Secrets.__instance = self
        else:
            # should anyone call Secrets directly, make a bold statement about it.
            raise Exception(
                """Secrets object already initialized - you cannot create another instance!
                (hint: use Secrets.instance()"""
            )
        # The vault client
        self.__client = None
        self.__cache = {}

    def connect(self):
        if self.__client is not None:
            return True

        vault_url = OSVars.get(VAULT_URL_KEY)

        print(f"connecting to vault on: {vault_url}")

        vault_user = OSVars.get(VAULT_USER_KEY)
        vault_pass = OSVars.get(VAULT_PASS_KEY)

        try:
            self.__client = hvac.Client(url=vault_url, timeout=30)
            self.__client.auth_userpass(vault_user, vault_pass)
        except Exception as ex:
            print(f"Failed connecting to vault. error: {ex}")
            return False

        print(f"connected to vault on {vault_url}")
        return True

    @staticmethod
    def get(path_to_secret):
        """
        Get the secret associated with the the provided key.
        Ex. Secrets.get("secret/common") for common secrets in vault

        Arguments:
            path {str} -- path to secret in vault

        Raises:
            Exception: if SM not connected
            Exception: if SM key isn't found

        Returns:
            [dict] -- the secret
        """
        # ensures connection
        Secrets.instance().connect()
        return Secrets.instance().__get(path_to_secret)

    def __get(self, path):
        """
        Get the secret associated with the the provided key.
        Ex. Secrets.get("secret/common") for common secrets in vault

        Arguments:
            path {str} -- path to secret in vault

        Raises:
            Exception: if SM not connected
            Exception: if SM key isn't found

        Returns:
            [dict] -- the secret
        """

        # try to hit cache first
        if path in self.__cache:
            return self.__cache[path]

        # else read remote
        if self.__client is None:
            raise Exception(
                """not connected to Secret-Management server.
                   Make sure to call connect first
                """
            )

        secret = self.__client.read(path)
        # using vault v2 the below is the new way to fetch a secret
        # secret = self.__client.secrets.kv.v2.read_secret_version(path)
        if secret is None or "data" not in secret:
            err = "Failed to read secret"
            print(err)
            raise Exception(err)

        self.__cache[path] = secret["data"]
        return secret["data"]

File: ruby/src/secrets.rb
require 'singleton'
require 'vault'
require_relative 'os_vars'
require_relative 'utils/logger'
require_relative 'utils/dict_utils'

module TwistConf
  #############################################################################
  # GLOBALS and CONSTANTS                                                     #
  #############################################################################

  # TODO: move to a common folder within the config clients monorepo
  VAULT_API_VERSION = 'v1'
  VAULT_URL_KEY = 'VAULT_URL'
  VAULT_DEFAULT_URL = 'https://vault.twistbioscience-staging.com'
  VAULT_USER_KEY = 'VAULT_USER'
  VAULT_PASS_KEY = 'VAULT_PASSWORD'
  ENVS_VAULT_KEY = 'envs'

  #############################################################################
  # IMPLEMENTATION                                                            #
  #############################################################################

  OSVars.register_mandatory(VAULT_USER_KEY, 'Vault secret management user name', String)
  OSVars.register_mandatory(VAULT_PASS_KEY, 'Vault secret management password', String)
  OSVars.register(VAULT_URL_KEY, 'Vault secret management server', String, VAULT_DEFAULT_URL)

  class Secrets
    include Singleton

    def initialize
      @__logged_in = false
      # secrets cache (key is category!)
      @__secrets = {}
      # secrets cache (key is path to secret)
      @__path_to_secret = {}
      @__vault = nil
    end

    # login to vault secret management server
    #
    # @return [void or throws]
    def login
      if @__logged_in
        return true
      end

      vault_url = OSVars.get(VAULT_URL_KEY)
      Log.info("connecting to vault via #{vault_url}")

      begin
        # TODO: might need to provide interface to services insisting on configuring vault client
        # Vault.configure do |config|
        #   config.timeout = 30
        #   config.ssl_timeout  = 5
        #   config.open_timeout = 5
        #   config.read_timeout = 30
        # end
        Vault.address = vault_url
        v_user = OSVars.get(VAULT_USER_KEY)
        v_pass = OSVars.get(VAULT_PASS_KEY)
        Vault.auth.userpass(v_user, v_pass)

        @__logged_in = true
      rescue StandardError => e
        raise "Failed connecting to vault on #{vault_url} error details: #{e}"
      end
    end

    def __perform_override(secret, path_to_secret)
      twist_env = OSVars.get('TWIST_ENV')
      result = nil

      if !secret[ENVS_VAULT_KEY.to_sym].nil? && !secret[ENVS_VAULT_KEY.to_sym][twist_env.to_sym].nil?
        Log.info("SECRET:: overriding env secret from #{path_to_secret}/#{ENVS_VAULT_KEY}/#{twist_env}")
        overriding = secret[ENVS_VAULT_KEY.to_sym][twist_env.to_sym]
        result = override_dict(secret, overriding)
        result.delete(ENVS_VAULT_KEY.to_sym)
      else
        result = secret
      end
      result
    end

    # ensuring secret exists + preload
    def require_secret(secret_category, path_to_secret)
      secret = get_by_path(path_to_secret)
      secret = Hash[__perform_override(secret, path_to_secret)]
      @__secrets[secret_category] = secret
      true
    rescue StandardError => e
      raise "Failed loading secret at #{secret_category}. Ex: #{e}"
    end

    # get a secret without specifying category - for non declarative purposes (it will be cached regardless)
    def get_by_path(path_to_secret)
      login
      # trying to hit internal cache
      return @__path_to_secret[path_to_secret] unless @__path_to_secret[path_to_secret].nil?

      secret_hash = nil
      # or read remote
      Vault.with_retries(Vault::HTTPConnectionError, Vault::HTTPError) do |attempt, e|
        # Note: Using puts since Rails logger won't yet be initialized
        raise "Failed fetching secret #{path_to_secret} from Vault. exception #{e} - attempt #{attempt}" if e

        Log.debug("Fetching secret from #{path_to_secret}")
        secret_hash = Vault.logical.read(path_to_secret).data
      end
      @__path_to_secret[path_to_secret] = secret_hash
      secret_hash
    end

    # helper static method for easier access (Secrets.get) without calling login
    def self.get(secret_category)
      Secrets.instance.login
      Secrets.instance.__get(secret_category)
    end

    # Main secret accessor. Uses internal cache (naive)
    #
    # @param secret_category [String]
    # @return [Hash or throws]
    def __get(secret_category)
      unless @__logged_in
        raise 'not connected to Secret-Management server.' \
              'Make sure to call connect first'
      end

      # category was not required and preloaded
      if @__secrets[secret_category].nil?
        raise "Unknown secret category [#{secret_category}]"
      end

      @__secrets[secret_category]
    end
  end
end

# see kitchen_sink.rb for better usage examples.
#
# OSVars.instance.init
# Secrets.instance.login
# sec = Secrets.get_by_path("secret/common")
# puts "common secret is... #{sec}"
